gh-122358: Remove re._compile

wimglenn opened the pull request:
<!-- gh-issue-number: gh-122358 -->
* Issue: gh-122358
<!-- /gh-issue-number -->

ZeroIntensity commented on Lib/re/__init__.py:
```diff
@@ -275,60 +275,17 @@ def findall(pattern, string, flags=0):
-    "Clear the regular expression caches"
-    _cache.clear()
-    _cache2.clear()
-    _compile_template.cache_clear()
-
-
-# SPECIAL_CHARS
-# closing ')', '}' and ']'
-# '-' (a range in character set)
-# '&', '~', (extended character set operations)
-# '#' (comment) and WHITESPACE (ignored) in verbose mode
-_special_chars_map = {i: '\\' + chr(i) for i in b'()[]{}?*+-|^$\\.&~# \t\n\r\v\f'}
-
-def escape(pattern):
-    """
-    Escape special characters in a string.
-    """
-    if isinstance(pattern, str):
-        return pattern.translate(_special_chars_map)
-    else:
-        pattern = str(pattern, 'latin1')
-        return pattern.translate(_special_chars_map).encode('latin1')
-
-Pattern = type(_compiler.compile('', 0))
-Match = type(_compiler.compile('', 0).match(''))
-
-# --------------------------------------------------------------------
-# internals
-
-# Use the fact that dict keeps the insertion order.
-# _cache2 uses the simple FIFO policy which has better latency.
-# _cache uses the LRU policy which has better hit rate.
-_cache = {}  # LRU
-_cache2 = {}  # FIFO
-_MAXCACHE = 512
-_MAXCACHE2 = 256
-assert _MAXCACHE2 < _MAXCACHE
-
-def _compile(pattern, flags):
-    # internal: compile pattern
```
Generally, it's a bad idea to refactor things unrelated to the issue.

wimglenn replied:
I did not refactor anything unrelated. This is just a poor github diff rendering.

ZeroIntensity replied:
Ah, I see that now. My bad!

eendebakpt replied:
Diff indeed looks larger than it is. By setting `_compile = compile` the diff can be made a bit smaller, but I doubt that is worth it

wimglenn committed:
Remove re._compile

Move implementation directly into re.compile.
Simplifies tracebacks and reduces call stack.

ZeroIntensity reviewed (approved):
LGTM, GitHub is making the diff look weird.

wimglenn committed:
Add news entry

AA-Turner commented on Misc/NEWS.d/next/Library/2024-08-04-13-22-32.gh-issue-122358.wpnaq4.rst:
```diff
@@ -0,0 +1 @@
+Remove re._compile, leaving a compatibility alias to re.compile.
```
```suggestion
Remove :func:`!re._compile`, leaving a compatibility alias to :func:`re.compile`.
```

AA-Turner commented on Lib/test/test_re.py:
```diff
@@ -1227,7 +1227,7 @@ def test_pickling(self):
             newpat = pickle.loads(pickled)
             self.assertEqual(newpat, oldpat)
-        # current pickle expects the _compile() reconstructor in re module
+        # previous pickles may expect the _compile() reconstructor in re module
```
Is it worth adding a test for pickles that expect ``_compile()``? ("No" is a valid answer)

wimglenn replied:
I've added it in https://github.com/python/cpython/pull/122357/commits/df79dd5cc1467f4cfc9e9902d216498e7391f51e

wimglenn committed:
Update Misc/NEWS.d/next/Library/2024-08-04-13-22-32.gh-issue-122358.wpnaq4.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

serhiy-storchaka committed:
Merge branch 'main' into re-compile

wimglenn committed:
reduce stacklevel in warnings

wimglenn committed:
Add a test for pickles that expect _compile()

wimglenn committed:
Skip re pkg frames in warnings

This uses the warnings feature skip_file_prefixes added in 3.12
https://github.com/python/cpython/pull/100840/

wimglenn committed:
Merge branch 'main' into re-compile

wimglenn committed:
Merge branch 'main' into re-compile

wimglenn committed:
Merge branch 'main' into re-compile

wimglenn committed:
Merge branch 'main' into re-compile

wimglenn committed:
Merge branch 'main' into re-compile

wimglenn committed:
Merge branch 'main' into re-compile

wimglenn committed:
Merge branch 'main' into re-compile
